robot/client/client_test.go

```diff
@@ -98,6 +98,117 @@ var finalResources = []resource.Name{
+}
+
+func (ms *mockRPCSubtypesUnimplemented) ResourceNames(
+	ctx context.Context, _ *pb.ResourceNamesRequest) (*pb.ResourceNamesResponse, error) {
+	board := board.Named("micro-rdk")
+	rNames := []*commonpb.ResourceName{
+		{
+			Namespace: string(board.API.Type.Namespace),
+			Type:      board.API.Type.Name,
+			Subtype:   board.API.SubtypeName,
+			Name:      board.Name,
+		},
+	}
+	return &pb.ResourceNamesResponse{Resources: rNames}, nil
+}
+
+type mockRPCSubtypesImplemented struct {
+	mockRPCSubtypesUnimplemented
+}
+
+func (ms *mockRPCSubtypesImplemented) ResourceRPCSubtypes(
+	ctx context.Context, _ *pb.ResourceRPCSubtypesRequest,
+) (*pb.ResourceRPCSubtypesResponse, error) {
+	return &pb.ResourceRPCSubtypesResponse{}, nil
+}
+
+func TestUnimplementedRPCSubtypes(t *testing.T) {
+	var client1 *RobotClient // test implemented
+	var client2 *RobotClient // test unimplemented
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*1)
+	defer cancel()
+
+	var port1 int
+	var port2 int
+	var err error
+	logger := logging.NewTestLogger(t)
+	for i := 0; i < 10; i++ {
+		port1, err = utils.TryReserveRandomPort()
+		if err != nil {
+			break
+		}
+	}
+	test.That(t, err, test.ShouldBeNil)
+	for i := 0; i < 10; i++ {
+		port2, err = utils.TryReserveRandomPort()
+		if err != nil {
+			break
+		}
+	}
+	test.That(t, err, test.ShouldBeNil)
+
+	listener1, err := net.Listen("tcp", fmt.Sprint("localhost:", port1))
+	test.That(t, err, test.ShouldBeNil)
+	listener2, err := net.Listen("tcp", fmt.Sprint("localhost:", port2))
+	test.That(t, err, test.ShouldBeNil)
+
+	rpcServer1, err := rpc.NewServer(logger.AsZap(), rpc.WithUnauthenticated())
+	test.That(t, err, test.ShouldBeNil)
+	defer rpcServer1.Stop()
+
+	rpcServer2, err := rpc.NewServer(logger.AsZap(), rpc.WithUnauthenticated())
+	test.That(t, err, test.ShouldBeNil)
+	defer rpcServer2.Stop()
+
+	err = rpcServer1.RegisterServiceServer(
+		ctx,
+		&pb.RobotService_ServiceDesc,
+		&mockRPCSubtypesImplemented{},
+		pb.RegisterRobotServiceHandlerFromEndpoint,
+	)
+	test.That(t, err, test.ShouldBeNil)
+
+	err = rpcServer2.RegisterServiceServer(
+		ctx,
+		&pb.RobotService_ServiceDesc,
+		&mockRPCSubtypesUnimplemented{},
+		pb.RegisterRobotServiceHandlerFromEndpoint,
+	)
+	test.That(t, err, test.ShouldBeNil)
+
+	go func() {
+		test.That(t, rpcServer1.Serve(listener1), test.ShouldBeNil)
+	}()
+
+	go func() {
+		test.That(t, rpcServer2.Serve(listener2), test.ShouldBeNil)
+	}()
+
+	client1, err = New(
+		ctx,
+		listener1.Addr().String(),
+		logger,
+	)
+	test.That(t, err, test.ShouldBeNil)
+	defer client1.Close(ctx)
+	test.That(t, client1.Connected(), test.ShouldBeTrue)
+	test.That(t, client1.rpcSubtypesUnimplemented, test.ShouldBeFalse)
+
+	client2, err = New(
+		ctx,
+		listener2.Addr().String(),
+		logger,
+	)
+	test.That(t, err, test.ShouldBeNil)
+	defer client2.Close(ctx)
+	test.That(t, client2.Connected(), test.ShouldBeTrue)
+	test.That(t, client2.rpcSubtypesUnimplemented, test.ShouldBeTrue)
+}
+
```

Comment:
I would like the 2nd test to test that after we get the unimplemented, we can update the server to return a different list of resource names, then we can refresh and verify that the client will have that different list of resource names while not calling the ResourceRPCSubtypes